File: algorithms/107. Binary Tree Level Order Traversal II/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        levels = []
        cur = [root] if root else []
        while cur:
            levels.append([node.val for node in cur])
            cur = [x for n in cur for x in (n.left, n.right) if x]
        return levels[::-1]

File: algorithms/107. Binary Tree Level Order Traversal II/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> result;
        if ( NULL == root ) {
            return result;
        }
        vector<TreeNode *> level(1, root);
        vector<TreeNode *> nextLevel;
        while ( level.size() > 0 ) {
            result.push_back(vector<int> ());
            for ( auto node : level ) {
                result.back().push_back(node->val);
                if ( node->left ) {
                    nextLevel.push_back(node->left);
                }
                if ( node->right ) {
                    nextLevel.push_back(node->right);
                }
            }
            level = nextLevel;
            nextLevel.clear();
        }
        reverse(result.begin(), result.end());
        return result;
    }
};

// 34 / 34 test cases passed.
// Status: Accepted
// Runtime: 3 ms
// beats 88.34%
